src/sentry/utils/circuit_breaker2.py

```
@@ -0,0 +1,185 @@
+        self.redis_pipeline = self.limiter.client.pipeline()
+
+        self.primary_quota = Quota(
+            self.window,
+            self.window_granularity,
+            self.error_limit,
+            f"{key}.circuit_breaker.ok",
+        )
+        self.recovery_quota = Quota(
+            self.window,
+            self.window_granularity,
+            self.recovery_error_limit,
+            f"{key}.circuit_breaker.recovery",
+        )
+
+        if self.recovery_error_limit >= self.error_limit:
+            self.recovery_error_limit = default_recovery_error_limit
+            logger.warning(
+                "Circuit breaker '%s' has a recovery error limit (%d) greater than or equal"
+                + " to its primary error limit (%d). Using the stricter error-limit-based"
+                + " default (%d) instead.",
+                key,
+                config["recovery_error_limit"],
+                self.error_limit,
+                self.recovery_error_limit,
+            )
+
+        # XXX: If we discover we have a config where we want this combo to work, we can consider
+        # using the `MockCircuitBreaker._clear_quota` helper, which is currently only used in tests,
+        # to clear out the main quota when we switch to the BROKEN state. (It will need tests of its
+        # own if so.)
+        if self.broken_state_duration + self.recovery_duration < self.window:
+            logger.warning(
+                "Circuit breaker '%s' has BROKEN and RECOVERY state durations (%d and %d sec, respectively)"
+                + " which together are less than the main error limit window (%d sec). This can lead to the"
+                + " breaker getting tripped unexpectedly, until the original spike in errors clears the"
+                + " main time window.",
+                key,
+                self.broken_state_duration,
+                self.recovery_duration,
+                self.window,
+            )
```

Rather than just a warning, IMO `if settings.DEBUG == True` (so in dev and CI) this should raise an exception

---

Oh, yeah - I can do that. Good idea!

---

UPDATE: Done.